academy/2-main-concepts/queries.md

```diff
@@ -9,34 +9,328 @@ tag: deep-dive
+    -proto ./proto/cosmos/staking/v1beta1/query.proto \  # Look into this .proto file for the Query protobuf service
+    -d '{"address":"$MY_DELEGATOR"}' \                   # Query arguments
+    localhost:9090 \                                     # gRPC server endpoint
+    cosmos.staking.v1beta1.Query/Delegations             # Fully-qualified service method name
+```
+
+## REST
+
+Users can also make queries through HTTP requests to a [REST server](https://docs.cosmos.network/master/core/grpc_rest.html#rest-server). The REST server is fully auto-generated from Protobuf services using gRPC-gateway.
+
+An example HTTP request for `MyQuery` would look like this:
+
+```
+GET http://localhost:1317/cosmos/staking/v1beta1/delegators/{delegatorAddr}/delegations
+```
+
+## How queries are handled by the CLI
+
+The examples above show how an external user can interact with a node by querying its state.
+
+To get a better understanding of the full lifecycle of a query, it helps to look at how the CLI prepares the query and how the node handles it in detail.
+
+From a user's perspective, the interactions are a bit different, but the underlying functions are almost identical because they are implementations of the same command defined by the module developer. This processing step takes place within the CLI, gRPC, or REST server and heavily involves a `client.Context`.
+
+## Context
+
+The first thing created in the execution of a CLI command is a `client.Context`. A `client.Context` is an object that stores all the data needed to process a request on the user side. It stores the following:
+
+* **Codec:** The [encoder/decoder](https://docs.cosmos.network/master/core/encoding.html) used by the application to marshal the parameters and query before making the Tendermint RPC request and unmarshal the returned response into a JSON object. The default codec used by the CLI is Protobuf.
+* **Account decoder:** The account decoder from the [auth](https://docs.cosmos.network/master/x/auth/spec/) module, which translates `[]bytes` into accounts.
+* **RPC client:** The Tendermint RPC client or node to which the request is relayed.
+* **Keyring:** A [key manager](https://docs.cosmos.network/master/basics/accounts.html#keyring) used to sign transactions and handle other operations with keys.
+* **Output writer:** A [writer](https://golang.org/pkg/io/#Writer) used to output the response.
+* **Configurations:** The flags configured by the user for this command, including `--height` specifying the height of the blockchain to query and `--indent` indicating adding an indent to the JSON response.
+
+The `client.Context` also contains various functions such as `Query()`, which retrieves the RPC Client and makes an ABCI call to relay a query to a full-node.
+
+```go
+// Context implements a typical context created in SDK modules for transaction
+// handling and queries.
+type Context struct {
```

Comment:
I think be up-to-date with the latest version; or maybe the used sdk version can be precised.